src/main/java/com/amazonaws/encryptionsdk/kms/StandardAwsKmsClientSuppliers.java

```diff
@@ -0,0 +1,252 @@
+        private static final Set<String> AWSKMS_METHODS = new HashSet<>();
+        private AWSKMSClientBuilder awsKmsClientBuilder;
+
+        static {
+            AWSKMS_METHODS.add("generateDataKey");
+            AWSKMS_METHODS.add("encrypt");
+            AWSKMS_METHODS.add("decrypt");
+        }
+
+        DefaultAwsKmsClientSupplierBuilder(AWSKMSClientBuilder awsKmsClientBuilder) {
+            this.awsKmsClientBuilder = awsKmsClientBuilder;
+        }
+
+        public AwsKmsClientSupplier build() {
+
+            return regionId -> {
+
+                if (clientsCache.containsKey(regionId)) {
+                    return clientsCache.get(regionId);
+                }
+
+                if (credentialsProvider != null) {
+                    awsKmsClientBuilder = awsKmsClientBuilder.withCredentials(credentialsProvider);
+                }
+
+                if (clientConfiguration != null) {
+                    awsKmsClientBuilder = awsKmsClientBuilder.withClientConfiguration(clientConfiguration);
+                }
+
+                if (regionId != null) {
+                    awsKmsClientBuilder = awsKmsClientBuilder.withRegion(regionId);
+                }
+
+                return newCachingProxy(awsKmsClientBuilder.build(), regionId);
+            };
+        }
+
+        /**
+         * Sets the AWSCredentialsProvider used by the client.
+         *
+         * @param credentialsProvider New AWSCredentialsProvider to use.
+         */
+        public DefaultAwsKmsClientSupplierBuilder credentialsProvider(AWSCredentialsProvider credentialsProvider) {
+            this.credentialsProvider = credentialsProvider;
+            return this;
+        }
+
+        /**
+         * Sets the ClientConfiguration to be used by the client.
+         *
+         * @param clientConfiguration Custom configuration to use.
+         */
+        public DefaultAwsKmsClientSupplierBuilder clientConfiguration(ClientConfiguration clientConfiguration) {
+            this.clientConfiguration = clientConfiguration;
+            return this;
+        }
+
+        /**
+         * Creates a proxy for the AWSKMS client that will populate the client into the client cache
+         * after an AWS KMS method successfully completes or an AWS KMS exception occurs. This is to prevent a
+         * a malicious user from causing a local resource DOS by sending ciphertext with a large number
+         * of spurious regions, thereby filling the cache with regions and exhausting resources.
+         *
+         * @param client   The client to proxy
+         * @param regionId The region the client is associated with
+         * @return The proxy
+         */
+        private AWSKMS newCachingProxy(AWSKMS client, String regionId) {
+            return (AWSKMS) Proxy.newProxyInstance(
+                    AWSKMS.class.getClassLoader(),
+                    new Class[]{AWSKMS.class},
+                    (proxy, method, methodArgs) -> {
+                        try {
```

Comment:
Is it worth remembering if we've successfully put it in the cache, and if so, short-circuit all of this logic and just delegate the call without the additional logic?

Reply:
I'm not sure I understand, the code already short-circuits on line 99 if the client is in the cache

Reply:
Yes. But for the first client supplied, a proxy is returned with this more expensive logic which is used every time.

Of course, if the KMS Key Ring always requests a new client for _every call_, then this is a non-issue.

Reply:
Got it. Yes the KMS Key ring requests a new client every call, it doesn't maintain any reference to the client it gets.